File: api/src/main/java/com/api/app/service/payment/strategy/PaymentGatewayFactory.java
package com.api.app.service.payment.strategy;

import com.api.app.emum.PAY005;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * PG사 전략 Factory
 * PAY005 enum의 가중치(referenceValue1)를 기반으로 PG사 선택
 *
 * @author Claude
 * @version 1.0
 * @since 2025-10-31
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class PaymentGatewayFactory {

    private final List<PaymentGatewayStrategy> strategies;
    private final Random random = new Random();

    /**
     * 가중치 기반 PG사 선택
     *
     * @return 선택된 PG 전략
     */
    public PaymentGatewayStrategy selectByWeight() {
        // PG 전략 맵 생성 (PAY005 -> strategy)
        Map<PAY005, PaymentGatewayStrategy> strategyMap = strategies.stream()
                .collect(Collectors.toMap(
                        PaymentGatewayStrategy::getPgType,
                        Function.identity()
                ));

        // 가중치 합계 계산
        int totalWeight = 0;
        for (PAY005 pg : PAY005.values()) {
            totalWeight += Integer.parseInt(pg.getReferenceValue1());
        }

        // 랜덤 선택
        int randomValue = random.nextInt(totalWeight) + 1;
        log.debug("PG selection - totalWeight={}, randomValue={}", totalWeight, randomValue);

        int currentWeight = 0;
        for (PAY005 pg : PAY005.values()) {
            currentWeight += Integer.parseInt(pg.getReferenceValue1());
            if (randomValue <= currentWeight) {
                log.info("PG selected: {} (weight={})", pg.getCodeName(), pg.getReferenceValue1());
                return strategyMap.get(pg);
            }
        }

        // 기본값: 첫 번째 전략
        log.warn("PG selection failed, using default strategy");
        return strategies.get(0);
    }

    /**
     * 코드로 PG 전략 조회
     *
     * @param pgTypeCode PG 타입 코드
     * @return PG 전략
     */
    public PaymentGatewayStrategy getByCode(String pgTypeCode) {
        PAY005 pg = PAY005.findByCode(pgTypeCode);
        return strategies.stream()
                .filter(s -> s.getPgType() == pg)
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("지원하지 않는 PG입니다: " + pgTypeCode));
    }

    /**
     * PAY005 enum으로 전략 조회
     *
     * @param pgType PG enum
     * @return PG 전략
     */
    public PaymentGatewayStrategy getStrategy(String pgTypeCode) {
        return strategies.stream()
                .filter(s -> s.getPgType().isEquals(pgTypeCode))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("지원하지 않는 PG입니다: " + pgTypeCode));
    }
}

File: api/src/main/kotlin/com/api/app/service/payment/strategy/PaymentGatewayFactory.kt
package com.api.app.service.payment.strategy

import com.api.app.emum.PAY005
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import kotlin.random.Random

@Component
class PaymentGatewayFactory(private val strategies: List<PaymentGatewayStrategy>) {

    private val log = LoggerFactory.getLogger(this::class.java)

    fun selectByWeight(): PaymentGatewayStrategy {
        val strategyMap = strategies.associateBy { it.getPgType() }
        val selectablePgs = PAY005.entries.filter { it.referenceValue1.toInt() > 0 && strategyMap.containsKey(it) }

        val totalWeight = selectablePgs.sumOf { it.referenceValue1.toInt() }
        if (totalWeight <= 0) {
            log.warn("No weighted PG strategy is registered, using default strategy")
            return strategies.first()
        }
        val randomValue = Random.nextInt(totalWeight) + 1

        var currentWeight = 0
        for (pg in selectablePgs) {
            currentWeight += pg.referenceValue1.toInt()
            if (randomValue <= currentWeight) {
                log.info("PG selected: {} (weight={})", pg.codeName, pg.referenceValue1)
                return strategyMap[pg] ?: strategies.first()
            }
        }

        log.warn("PG selection failed, using default strategy")
        return strategies.first()
    }

    fun getStrategy(pgType: PAY005): PaymentGatewayStrategy =
        strategies.firstOrNull { it.getPgType() == pgType }
            ?: throw IllegalArgumentException("지원하지 않는 PG입니다: ${pgType.codeName}")

    fun getByCode(pgTypeCode: String): PaymentGatewayStrategy {
        val pg = PAY005.findByCode(pgTypeCode)
            ?: throw IllegalArgumentException("지원하지 않는 PG 코드입니다: $pgTypeCode")
        return getStrategy(pg)
    }
}
